File: Java_Virtual_World/src/Gameplay/Organism.java
package Gameplay;
import java.awt.*;
import java.util.Vector;

abstract public class Organism {
    protected int x;
    protected int y;
    protected int strength;
    protected int initiative;
    protected int age = 0;
    protected boolean alive = true;
    protected World world;

    protected Organism(int x, int y, int strength, int initiative) {
        this.x = x;
        this.y = y;
        this.strength = strength;
        this.initiative = initiative;
    }

    //Setters
    public void setX(int x){
        this.x = x;
    }
    public void setY(int y){
        this.y = y;
    }
    public void setStrength(int s){
        strength = s;
    }
    public void setInitiative(int i){
        initiative = i;
    }
    public void setAge(int a){
        age = a;
    }
    public void setWorld(World w){
        world = w;
    }

    //Getters
    public int getX(){
        return x;
    }
    public int getY(){
        return y;
    }
    public int getStrength(){
        return strength;
    }
    public int getInitiative(){
        return initiative;
    }
    public int getAge(){
        return age;
    }
    public boolean isAlive(){
        return alive;
    }

    public void kill(){
    world.getNots().addNode(toString() + "is dead");
    alive = false;
    }

    public boolean didBlocked(Organism opponent){
        return false;
    }
    public boolean isSmelling(){
        return false;
    }
    public boolean didEscape(){
        return false;
    }
    public boolean escape(){
        if(didEscape()){
            Vector<Integer> tmp = world.getFreeSquare(x, y);

            if(!(tmp.get(0) == x && tmp.get(1) == y)){
                x = tmp.get(0);
                y = tmp.get(1);
                return true;
            }
        }
        return false;
    }
    public void nextYear(){
        age++;
    }

    public abstract void Action();
    public abstract void Collision();
    public abstract String Draw();
    public abstract void NextRound();
    public abstract Organism Copy();
    @Override
    abstract public String toString();
}

File: Python_Virtual_World/Organism.py
from abc import ABC, abstractmethod

from MyVector import MyVector


class Organism(ABC):


    def __init__(self, position : MyVector, s : int, i : int):

        self._position = position
        self._strength = s
        self._initiative = i
        self._alive = True
        self._age = 0



    def getPosition(self):

        return self._position


    def setPosition(self, p : MyVector):

        self._position = p


    def getInitiative(self):

        return self._initiative


    def getStrength(self):

        return self._strength


    def setStrength(self, s : int):

        self._strength = s


    def getAge(self):

        return self._age

    def setAge(self, a: int):

        self._age = a
    

    def setWorld(self, w):

        self._world = w


    def isAlive(self) -> bool:

        return self._alive


    def kill(self):

        self._world.getNots().add(f"{str(self)} is dead")
        self._alive = False


    def giveMod(self, other):

        pass


    def didBlocked(self,other) -> bool:

        return False

    def isResistance(self) -> bool:

        return False


    def didEscape(self) -> bool:

        return False

    def escape(self) -> bool:

        if self.didEscape():

            newField = self._world.getFreeSquare(self._position)

            if newField == self._position:

                return False

            self.setPosition(newField)

            return True

        return False



    def isSmelling(self):

        return False


    def ageUp(self):

        self._age+=1


    @abstractmethod
    def __str__(self):
        pass


    @abstractmethod
    def action(self):

        pass


    @abstractmethod
    def collision(self):

        pass


    @abstractmethod
    def draw(self) -> str:

        pass


    @abstractmethod
    def newRound(self):

        pass




